sled-agent/src/vmm_reservoir.rs

```diff
@@ -0,0 +1,272 @@
+        match (percentage, size_mb) {
+            (None, None) => None,
+            (Some(p), None) => Some(ReservoirMode::Percentage(p)),
+            (None, Some(mb)) => Some(ReservoirMode::Size(mb)),
+            (Some(_), Some(_)) => panic!(
+                "only one of vmm_reservoir_percentage and \
+                vmm_reservoir_size_mb is allowed"
+            ),
+        }
+    }
+}
+
+/// A message sent from [`VmmReservoirManagerHandle`] to [`VmmReservoirManager`]
+enum ReservoirManagerMsg {
+    SetReservoirSize {
+        mode: ReservoirMode,
+        reply_tx: oneshot::Sender<Result<(), Error>>,
+    },
+}
+
+#[derive(Clone)]
+/// A mechanism to interact with the [`VmmReservoirManager`]
+pub struct VmmReservoirManagerHandle {
+    reservoir_size: Arc<AtomicU64>,
+    tx: flume::Sender<ReservoirManagerMsg>,
+    // A notification channel indicating that the size of the VMM reservoir has
+    // changed. We use a broadcast channel instead of a `Notify` to prevent lost
+    // updates with multiple receivers. Importantly, a `RecvError::Lagged` is
+    // just as valuable as an `Ok(())`, and so this acts as a pure notification
+    // channel.
+    size_updated_tx: broadcast::Sender<()>,
+    _manager_handle: Arc<thread::JoinHandle<()>>,
+}
+
+impl VmmReservoirManagerHandle {
+    /// Returns the last-set size of the reservoir
+    pub fn reservoir_size(&self) -> ByteCount {
+        self.reservoir_size.load(Ordering::SeqCst).try_into().unwrap()
+    }
+
+    // Return the receiver that notifies us about a size update. The value
+    // itself is held in an atomic. While we could have replaced both of these
+    // with a `watch`, that is not the semantics all callers want. Sometimes
+    // they just want to "read" the latest size of the reservoir.
+    pub fn subscribe_for_size_updates(&self) -> broadcast::Receiver<()> {
+        self.size_updated_tx.subscribe()
+    }
+
+    /// Tell the [`VmmReservoirManager`] to set the reservoir size and wait for
+    /// a response.
+    ///
+    /// Returns an error if the allocation fails or if the manager is currently
+    /// busy handling another request.
+    //
+    // It's anticipated this will be used to change the reservation
+    #[allow(unused)]
+    pub async fn set_reservoir_size(
+        &self,
+        mode: ReservoirMode,
+    ) -> Result<(), Error> {
+        let (tx, rx) = oneshot::channel();
+        let msg = ReservoirManagerMsg::SetReservoirSize { mode, reply_tx: tx };
+        if let Err(e) = self.tx.try_send(msg) {
+            return Err(match e {
+                flume::TrySendError::Full(e) => Error::Busy,
+                flume::TrySendError::Disconnected(e) => Error::Shutdown,
+            });
+        }
+        rx.await.map_err(|_| Error::ReplySenderDropped)?
+    }
+}
+
+/// Manage the VMM reservoir in a background thread
+pub struct VmmReservoirManager {
+    reservoir_size: Arc<AtomicU64>,
+    rx: flume::Receiver<ReservoirManagerMsg>,
```

Comment:
I'm probably just missing something, but why does this need to be a `flume` channel? The only consumer of these messages I can see is the task, so this seems like it could just be a regular MPSC channel.

Reply:
It doesn't. It could just as easily be an mpsc, but we already use flume elsewhere and it's supposedly faster. Why not make everything go zoom for no reason!